### DISCUSSION TITLE
Executing Action/Goal over Roslib.js (WebSockets)

### DISCUSSION THREAD

**Discussion by Pankaj2101** [iRobotEducation/create3_docs#322] [Questions]
I use the following command to rotate the Create3 robot.
_**ros2 action send_goal /rotate_angle irobot_create_msgs/action/RotateAngle "{angle: 1.57,max_rotation_speed: 0.5}"**_
This is working perfectly well through the terminal.

Now I want to execute the same through a Roslib.js (i.e. Websockets) for which I wrote the following web page.

```
**_<html>

<head>
    <meta charset="utf-8" />
    <style>
        body
        {
            width: 600px;
            font-family: calibri;
        }

        #LogWindow
        {
            cursor: default;
            min-height: 300px;
            overflow: auto;
            background: #fff;
            border: 1px solid #ffaaaa;
            border-radius: 4px;
            color: #ff5555;
            margin: 3px 0;
            padding: 5px;
            cursor: pointer;
        }
    </style>
    <script src="./Scripts/jquery-3.6.3.min.js"></script>
    <script src="./Scripts/eventemitter2.min.js"></script>
    <script src="./Scripts/roslib.js"></script>
    <script>
        $(function ()
        {
            RotateRobot();
        });

        function RotateRobot()
        {
            var ObjRos = new ROSLIB.Ros();

            // Initialise the RosLib
            ObjRos.connect("ws://192.168.0.201:9090");
            console.log("Library Initialised");

            // Handle Connection and Error events.
            ObjRos.on("connection", function ()
            {
                ShowMessage("Connected!");
            });

            ObjRos.on("close", function ()
            {
                ShowMessage("Disconnected!");
            });

            ObjRos.on("error", function (Error)
            {
                ShowMessage("Error: " + JSON.stringify(Error));
            });

            // Exceute Action
            var ObjActionClient;
            var ObjGoal; 

            //ros2 action send_goal /rotate_angle irobot_create_msgs/action/RotateAngle "{angle: 1.57,max_rotation_speed: 0.5}"
            ObjActionClient = new ROSLIB.ActionClient(
                {
                    ros: ObjRos,
                    serverName: "/rotate_angle",
                    actionName: "irobot_create_msgs/RotateAngle"
                });

            // Create a goal.
            ObjGoal = new ROSLIB.Goal(
                {
                    actionClient: ObjActionClient,
                    goalMessage: { angle: 1.57, max_rotation_speed: 0.5 }
                });

            // Print out their output into the terminal.
            ObjGoal.on('feedback', function (feedback)
            {
                ShowMessage('Feedback: ' + JSON.stringify(feedback));
            });

            ObjGoal.on('result', function (result)
            {
                ShowMessage('Final Result: ' + JSON.stringify(result));
            });

            // Send the goal to the action server.
            ObjGoal.send();
        };

        function ShowMessage(Message)
        {
            $('#LogWindow').append("<div>" + Message + "</div>");
        }
    </script>
</head>

<body>
    <h1>RosLib Action Test</h1>
    <div id="LogWindow">
        <div class="EventInfo">Connecting to RosBridge...</div>
    </div>
</body>

</html>_**

```
**_The terminal for websocket shows the following error.
td-otis@td-otis:~$ ros2 launch rosbridge_server rosbridge_websocket_launch.xml
[INFO] [launch]: All log files can be found below /home/td-otis/.ros/log/2023-01-28-00-00-33-718123-td-otis-4856
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rosbridge_websocket-1]: process started with pid [4858]
[INFO] [rosapi_node-2]: process started with pid [4860]
[rosbridge_websocket-1] [INFO] [1674844234.583450344] [rosbridge_websocket]: Rosbridge WebSocket server started on port 9090

[rosbridge_websocket-1] [INFO] [1674844245.684127991] [rosbridge_websocket]: Client connected. 1 clients total.
[rosbridge_websocket-1] [ERROR] [1674844245.749636686] [rosbridge_websocket]: [Client 13b1e924-ee4e-4428-ac54-fc2ec7648c3c] [id: advertise:/rotate_angle/goal:1] advertise: Unable to import msg class RotateAngleGoal from package irobot_create_msgs. Caused by module 'irobot_create_msgs.msg' has no attribute 'RotateAngleGoal'
[rosbridge_websocket-1] [INFO] [1674844245.757314874] [rosbridge_websocket]: [Client 13b1e924-ee4e-4428-ac54-fc2ec7648c3c] Subscribed to /rotate_angle/status
[rosbridge_websocket-1] [ERROR] [1674844245.758339606] [rosbridge_websocket]: [Client 13b1e924-ee4e-4428-ac54-fc2ec7648c3c] [id: subscribe:/rotate_angle/feedback:4] subscribe: Unable to import msg class RotateAngleFeedback from package irobot_create_msgs. Caused by module 'irobot_create_msgs.msg' has no attribute 'RotateAngleFeedback'
[rosbridge_websocket-1] [ERROR] [1674844245.759216402] [rosbridge_websocket]: [Client 13b1e924-ee4e-4428-ac54-fc2ec7648c3c] [id: subscribe:/rotate_angle/result:5] subscribe: Unable to import msg class RotateAngleResult from package irobot_create_msgs. Caused by module 'irobot_create_msgs.msg' has no attribute 'RotateAngleResult'
[rosbridge_websocket-1] [ERROR] [1674844245.759969995] [rosbridge_websocket]: [Client 13b1e924-ee4e-4428-ac54-fc2ec7648c3c] [id: publish:/rotate_angle/goal:6] publish: Cannot infer topic type for topic /rotate_angle/goal as it is not yet advertised_**

Can someone point out the correct code/servername/actionname etc?

**Comment by justinIRBT** [comment]
I haven't used Roslib.js or rosbridge with actions, but the python syntax it is trying to use seems off.  For example, the `RotateAngle` class comes from `irobot_create_msgs.action` not `irobot_create_msgs.msg`.  An example of using the RotateAngle action in python is:
```
from irobot_create_msgs.action import RotateAngle
goal = RotateAngle.Goal()
```
I am not sure what is making the rosbridge look for `'irobot_create_msgs.msg.RotateAngleGoal'`, but that isn't the correct syntax.

**Reply by justinIRBT** [reply]
I see someone with a similar issue here, but no resolution in the thread:
https://answers.ros.org/question/404992/rosbridge_websocket-unable-to-import/

**Reply by Pankaj2101** [reply]
Does it mean no one has ever tested actions over Roslib?
Ok. A simple question - do you think the problem is in roslib.us or the rosbridge or in the create 3 msgs?
I am badly stuck I HAVE to call create 3 over websockets ONLY. 
If any resource who can help, please let me know

**Comment by justinIRBT** [comment]
Given the linked ROS issue thread, the issue appears to be a general issue with using ROS lib.js with ros2 actions, not specific to create3.  What is driving your requirement to use websockets?  If you just need to use a web page, could you write a python webserver that could invoke the actions with rclpy?

**Reply by Pankaj2101** [reply]
That can solve my problem as well and I am perfectly ok with a Python web server.

But what I read is that it needs to be compiled as a node and cannot be run from a Python web server.  (Read in an article "https://jimbobbennett.dev/blogs/irobot-create3-subscribe-to-messages/"  which says, "..rclpy is the ROS Python library and has all the functionality you need to interact with a robot. But there is a catch - you can’t just create a Python script and use rclpy, you have to build a ROS2 application inside a ROS2 workspace...").

Can you point/give me the smallest Python webserver template which can execute one ACTION say "Rotate Robot", without compiling the python script each time I change it?

My need for websocket is for this very reason. I need to dynamically change the actions based on a web or desktop user interface.

**Reply by justinIRBT** [reply]
I don’t have an example with a webserver calling an action, but you can see an example of using a python rclpy application with create3 here https://github.com/iRobotEducation/create3_examples
See
https://github.com/iRobotEducation/create3_examples/tree/galactic/create3_examples_py
You just need to colcon build a workspace with the instructions in create3_examples.  If you are running rosbridge on your system, your system should be capable of building and running the example.  You can use that as a starting point for a python rclpy project and build your webserver with action into that.  There are instructions on writing a python action client here https://docs.ros.org/en/foxy/Tutorials/Intermediate/Writing-an-Action-Server-Client/Py.html
